libs/zpool/src/handle.zig

```diff
@@ -57,78 +57,71 @@ pub fn Handle(
+            bits: packed struct {
+                index: CompactIndex,
+                cycle: CompactCycle,
+            },
         };
 
+        pub const Resource = TResource;
+
         pub const AddressableIndex = AddressableUInt(index_bits);
         pub const AddressableCycle = AddressableUInt(cycle_bits);
 
-        pub const max_index = ~@as(Index, 0);
-        pub const max_cycle = ~@as(Cycle, 0);
+        pub const max_index = ~@as(CompactIndex, 0);
+        pub const max_cycle = ~@as(CompactCycle, 0);
         pub const max_count = @as(Id, max_index - 1) + 2;
 
         id: Id = 0,
 
         pub const nil = Self{ .id = 0 };
 
-        pub fn init(_index: Index, _cycle: Cycle) Self {
-            var u = Union{ .handle = .{ .index = _index, .cycle = _cycle }};
+        pub fn init(index: CompactIndex, cycle: CompactCycle) Self {
+            var u = CompactUnion{ .bits = .{ .index = index, .cycle = cycle } };
             return .{ .id = u.id };
         }
 
+        /// Unpacks the `index` and `cycle` bit fields that comprise
+        /// `Handle.id` into an `AddressableHandle`, which stores
+        /// the `index` and `cycle` values in pointer-addressable fields.
         pub fn addressable(self: Self) AddressableHandle {
+            var u = CompactUnion{ .id = self.id };
             return .{
-                .index = self.index(),
-                .cycle = self.cycle(),
+                .index = u.bits.index,
+                .cycle = u.bits.cycle,
             };
         }
 
-        pub fn index(self: Self) AddressableIndex {
-            var u = Union{ .id = self.id };
-            return u.handle.index;
-        }
-
-        pub fn cycle(self: Self) AddressableCycle {
-            var u = Union{ .id = self.id };
-            return u.handle.cycle;
-        }
-
-        pub fn eql(a: Self, b: Self) bool {
-            return a.id == b.id;
-        }
-
+        /// When you want to directly access the `index` and `cycle` of a
+        /// handle, first convert it to an `AddressableHandle` by calling
+        /// `Handle.addressable()`.
+        /// An `AddressableHandle` can be converted back into a "compact"
+        /// `Handle` by calling `AddressableHandle.compact()`.
         pub const AddressableHandle = struct {
-            pub const Compact = Self;
-
             index: AddressableIndex = 0,
             cycle: AddressableCycle = 0,
 
-            pub fn compact(handle: AddressableHandle) Compact {
-                const _index = @intCast(Index, handle.index);
-                const _cycle = @intCast(Cycle, handle.cycle);
-                return Compact.init(_index, _cycle);
-            }
-
-            pub fn eql(a: AddressableHandle, b: AddressableHandle) bool {
-                return a.index == b.index and a.cycle == b.cycle;
+            /// Returns the corresponding `Handle`
+            pub fn compact(handle: AddressableHandle) CompactHandle {
```

Comment:
Perhaps I should rename this method simply `handle()`

Reply:
Yes, I think we should go with handle().